### ISSUE TITLE
trial option to exit on failure

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4634]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#1518|
|Type|enhancement|
|Created|2006-03-01 05:54:44Z|
|Branch|https://github.com/twisted/twisted/tree/exitfirst-1518|

Add an option to Trial to make it stop the test run on the first error/failure.

py.test uses the option:
```
    -x, --exitfirst    exit instantly on first error or failed test.
```
which looks fine to me.

Attachments:

* [exitfirst.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c8e/c8ec393b52a33555bfca80d15909f59f2d09d0cb/b22ed114c5489cfba641679e54af5872c9fec0e3.patch) (11283 bytes) - added by Julian on 2013-04-16 14:10:40Z - 
* [exitfirst.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c8e/c8ec393b52a33555bfca80d15909f59f2d09d0cb/bc33f8af9bf4bc77fb272a946b50901361e784df.patch) (4428 bytes) - added by Julian on 2013-05-23 14:26:19Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1518 1518
type__enhancement enhancement
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_exitfirst_1518 branches/exitfirst-1518
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1141192484000000 1141192484000000
changetime__1371241709000000 1371241709000000
version__ 
owner__Julian Julian
cc__Jerub cc__acapnotic cc__strank cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `new`|
|-|-|

Unassigning. Too many other important tickets.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_9"></a>|@Julian commented|
|-|-|

Up for review.

Left out support for `--jobs`, since that apparently is non-trivial though I didn't look into it. Filed as [#6436](https://github.com/twisted/twisted/issues/6436).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

(In [38448]) Branching to 'exitfirst-1518'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

(In [38449]) Commit exitfirst.patch refs [#4634](https://github.com/twisted/twisted/issues/4634)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

Here are some [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/exitfirst-1518).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set owner to @Julian|
|-|-|

Thanks for the contribution, Julian.

    1. The man page and the command line documentation really need to explain that the option is mutually exclusive with `--jobs`.
    2. I don't think that unexpected successes should cause the tests to exit.  In fact just a couple of hours ago I was remarking to someone that the `SUCCESS` message really ought to be yellow rather than red, since it doesn't cause a non-zero exit status (and therefore is not really a "failure").
    3. The blank line after the docstring but before the implementation of each test case looks superfluous to me.
    4. The NEWS file is no longer accurate; the flag isn't called `--stop` or `--stop-failed`.
    5. [http://buildbot.twistedmatrix.com/builders/twistedchecker/builds/678/steps/run-twistedchecker/logs/new%20twistedchecker%20errors There are some twistedchecker errors].  However, the way to fix this, I think, is to simply remove the docstrings for these implementation methods, as was [ticket:6459#comment:11 recently suggested to me].

Overall a pretty solid feature addition though; if you address these minor issues I think we can land it.  I've noticed your name quite a few times around here recently, and you seem to be following coding standards, testing standards etc pretty closely.  Would you like to become a committer?

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_14"></a>|@Julian removed owner|
|-|-|

Hey.

 1. Added.
 2. OK. I don't necessarily like that behavior personally, I use `--exitfirst` to be "hey there's a bug, go fix it"  most of the time, but I see py.test doesn't stop after `unexpectedSuccess`, and `nose` also probably doesn't but I can't tell at the moment because apparently it's using a `TestResult` that doesn't have `addUnexpectedSuccess` anyhow. So I'll cave :).
 3. OK.
 4. Fixed.
 5. Done

> Overall a pretty solid feature addition though; if you address these minor issues I think we can land it. I've noticed your name quite a few times around here recently, and you seem to be following coding standards, testing standards etc pretty closely. Would you like to become a committer?

Sure, thanks :) Though I doubt I'd actually help commit anything anytime soon until I spend some time figuring out / bugging tomprince to teach me how to use `git-svn` properly -- last I checked I think the Wiki needed some updating there now that things are moving along.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

Replying to [Julian](#note_14):
> Hey.
> 
>  1. Added.
>  2. OK. I don't necessarily like that behavior personally, I use `--exitfirst` to be "hey there's a bug, go fix it"  most of the time, but I see py.test doesn't stop after `unexpectedSuccess`, and `nose` also probably doesn't but I can't tell at the moment because apparently it's using a `TestResult` that doesn't have `addUnexpectedSuccess` anyhow. So I'll cave :).
>  3. OK.
>  4. Fixed.
>  5. Done

Thanks, I hope someone can take a look at this soon.

> > Overall a pretty solid feature addition though; if you address these minor issues I think we can land it. I've noticed your name quite a few times around here recently, and you seem to be following coding standards, testing standards etc pretty closely. Would you like to become a committer?
> 
> Sure, thanks :) Though I doubt I'd actually help commit anything anytime soon until I spend some time figuring out / bugging tomprince to teach me how to use `git-svn` properly -- last I checked I think the Wiki needed some updating there now that things are moving along.

The [git mirror documentation page](https://twistedmatrix.com/trac/wiki/GitMirror) is actually almost complete.  I can help you out with specifics if you have trouble.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_16"></a>|@tomprince commented|
|-|-|

(In [38690]) Apply exitfirst.2.patch from Julian.

Refs: [#4634](https://github.com/twisted/twisted/issues/4634)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_17"></a>|@tomprince set owner to @Julian|
|-|-|

>> 1. The man page and the command line documentation really need to explain that the option is mutually exclusive with `--jobs`.
> Added
The manpage has been updated, but the command-line help hasn't.

Other than that, this looks good. Please (get commit access and) merge.

**Comment by twisted-trac** [comment]
|<img alt="julian's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|julian set status to `closed`|
|-|-|

(In [38805]) Merge exitfirst-1518: Add --exitfirst to trial

Author: Julian
Reviewers: glyph, tomprince
Fixes: [#4634](https://github.com/twisted/twisted/issues/4634)

Added the --exitfirst flag, allowing one to stop a trial run after the first
non-successful test.
